adapter/stdio/stdio.go

```diff
@@ -0,0 +1,205 @@
+	}
+
+	zapBuilderFn func(outputPaths ...string) (*zap.Logger, error)
+)
+
+const (
+	stdErr = "stderr"
+	stdOut = "stdout"
+)
+
+// ensure our types implement the requisite interfaces
+var _ logentry.HandlerBuilder = builder{}
+var _ logentry.Handler = &handler{}
+var _ metric.HandlerBuilder = builder{}
+var _ metric.Handler = &handler{}
+
+///////////////// Configuration Methods ///////////////
+
+func (builder) Build(cfg proto.Message, _ adapter.Env) (adapter.Handler, error) {
+	c := cfg.(*config.Params)
+
+	outputPath := stdErr
+	if c.LogStream == config.STDOUT {
+		outputPath = stdOut
+	}
+
+	buildZap := newZapLogger
+
+	zapLogger, err := buildZap(outputPath)
+	if err != nil {
+		return nil, fmt.Errorf("could not build logger: %v", err)
+	}
+
+	sl := make(map[string]zapcore.Level)
+	for k, v := range c.SeverityLevels {
+		sl[k] = mapConfigLevel(v)
+	}
+
+	return &handler{
+		omitEmpty:      c.OmitEmptyFields,
+		severityLevels: sl,
+		metricLevel:    mapConfigLevel(c.MetricLevel),
+		zapper:         zapLogger,
+	}, nil
+}
+
+func mapConfigLevel(l config.Params_Level) zapcore.Level {
+	if l == config.WARNING {
+		return zapcore.WarnLevel
+	} else if l == config.ERROR {
+		return zapcore.ErrorLevel
+	}
+	return zapcore.InfoLevel
+}
+
+func (builder) ConfigureLogEntryHandler(map[string]*logentry.Type) error {
+	return nil
+}
+
+func (builder) ConfigureMetricHandler(map[string]*metric.Type) error {
+	return nil
+}
+
+func newZapLogger(outputPaths ...string) (*zap.Logger, error) {
+	prodConfig := zap.NewProductionConfig()
+	prodConfig.DisableCaller = true
+	prodConfig.DisableStacktrace = true
+	prodConfig.OutputPaths = outputPaths
+	prodConfig.EncoderConfig = zapcore.EncoderConfig{
+		LineEnding:     zapcore.DefaultLineEnding,
+		EncodeTime:     zapcore.ISO8601TimeEncoder,
+		EncodeDuration: zapcore.SecondsDurationEncoder,
+	}
+
+	return prodConfig.Build()
+}
+
+////////////////// Runtime Methods //////////////////////////
+
+func (h *handler) HandleLogEntry(_ context.Context, instances []*logentry.Instance) error {
+	var errors *multierror.Error
+
+	fields := make([]zapcore.Field, 0, 6)
+	for _, instance := range instances {
+		entry := zapcore.Entry{
+			LoggerName: instance.Name,
+			Level:      h.mapSeverityLevel(instance.Severity),
+
+			// TODO: disabled in template
+			// Time: instance.Timestemp
+		}
+
+		for k, v := range instance.Variables {
+			if v == "" && h.omitEmpty {
+				continue
+			}
+			fields = append(fields, zap.Any(k, v))
+		}
+
+		if err := h.zapper.Core().Write(entry, fields); err != nil {
+			errors = multierror.Append(errors, err)
+		}
+		fields = fields[:0]
+	}
+
+	return errors.ErrorOrNil()
+}
+
+func (h *handler) HandleMetric(_ context.Context, instances []*metric.Instance) error {
+	var errors *multierror.Error
+
+	fields := make([]zapcore.Field, 0, 6)
+	for _, instance := range instances {
+		entry := zapcore.Entry{
+			LoggerName: instance.Name,
+			Level:      h.metricLevel,
+			Time:       time.Now(),
```

Comment:
Arguably we should use `request.time` for this, but it's not easy to plumb the attribute through. We should probably promote it up to a field on the metric Instance.